### PR TITLE
[WIP]: chore(gatsby): migrate webpack utils to TypeScript

### PR DESCRIPTION
**Pull request by sasurau4**
## Description

Part of #21995 

I migrated `webpack-utils.js` to TypeScript

## Related Issues

#21995

### CODE REVIEW

File: packages/gatsby/src/utils/webpack-plugins.ts [hunk 1]
```diff
@@ -1,6 +1,9 @@
 import webpack, { Plugin } from "webpack"
 
-const plugin = (name: string, optimize?: boolean): Plugin => {
+const plugin = (
+  name: string,
+  optimize?: boolean
+): ((...args: any) => Plugin) => {
```
Comment (blainekasten) [inline packages/gatsby/src/utils/webpack-plugins.ts]: ha good catch. Thanks for fixing!

File: packages/gatsby/src/utils/webpack-utils.ts [hunk 2]
```diff
@@ -43,84 +31,100 @@ type Stage = "develop" | "develop-html" | "build-javascript" | "build-html"
...
-module.exports = async ({
+export const createUtils = async ({
   stage,
   program,
 }: {
-  stage: Stage,
-  program: any,
-}): Promise<WebpackUtilsOptions> => {
+  stage: Stage
+  program: any
```
Comment (blainekasten) [inline packages/gatsby/src/utils/webpack-utils.ts]: you can import `IProgram` from `src/commands/types`

File: packages/gatsby/src/utils/webpack-utils.ts [hunk 3]
```diff
@@ -43,84 +31,100 @@ type Stage = "develop" | "develop-html" | "build-javascript" | "build-html"
...
+}): Promise<IWebpackUtils> => {
```
Comment (blainekasten) [inline packages/gatsby/src/utils/webpack-utils.ts]: so I looked into this, there is no asynchroneous code in this function. We can drop the async and the Promise type

**Comment by blainekasten** [comment]
There is a large amount of `any` types in this file. Would it be possible to improve those types?

**Comment by sasurau4** [comment]
@blainekasten I'll try to reduce any as much as possible. :+1: 
Some other files need to migrate, so I'll try to migrate them at first.

**Pull request merged**